src/v/kafka/security/scram_algorithm.cc

```diff
@@ -0,0 +1,205 @@
+#include "kafka/security/scram_algorithm.h"
+
+#include "utils/base64.h"
+#include "utils/to_string.h"
+#include "utils/utf8.h"
+
+#include <boost/algorithm/string.hpp>
+
+namespace kafka {
+
+// ALPHA / DIGIT / "/" / "+"
+// NOLINTNEXTLINE
+#define BASE64_CHAR "[a-zA-Z0-9/+]"
+
+// base64-4 = 4base64-char
+// base64-3 = 3base64-char "="
+// base64-2 = 2base64-char "=="
+// base64   = *base64-4 [base64-3 / base64-2]
+// NOLINTNEXTLINE
+#define BASE64                                                                 \
+    "(?:" BASE64_CHAR "{4})*(?:" BASE64_CHAR "{3}=|" BASE64_CHAR "{2}==)?"
+
+// Printable ASCII except ","
+// NOLINTNEXTLINE
+#define PRINTABLE "[\\x21-\\x2b\\x2d-\\x7e]+"
+
+// NOLINTNEXTLINE
+#define EXTENSIONS "(,[a-zA-Z]+=[\\x01-\\x2b\\x2d-\\x7f]+)*"
+
+// UTF8-char except NUL, "=", and ","
+// NOLINTNEXTLINE
+#define VALUE_SAFE_CHAR "[\\x01-\\x2b\\x2d-\\x3c\\x3e-\\x7f]"
+
+// 1*(value-safe-char / "=2C" / "=3D")
+// NOLINTNEXTLINE
+#define SASLNAME "(?:" VALUE_SAFE_CHAR "|=2C|=3D)+"
+
+// value-char = value-safe-char / "="
+// value = 1*value-char
+// NOLINTNEXTLINE
+#define VALUE "[\\x01-\\x2b\\x2d-\\x7f]+"
+
+// NOLINTNEXTLINE
+#define ALPHA "[a-zA-Z]"
+
+// clang-format off
+static constexpr const char* client_first_message_re =
+    "n,(?:a=(" SASLNAME "))?,(m=" VALUE ",)?n=(" SASLNAME "),"
+    "r=(" PRINTABLE ")(," ALPHA "+=" VALUE ")*";
+
+static constexpr const char* client_final_message_re =
+    "c=(" BASE64 "),r=(" PRINTABLE ")" EXTENSIONS ",p=(" BASE64 ")";
+// clang-format on
+
+std::ostream& operator<<(std::ostream& os, const scram_credential& cred) {
+    fmt::print(
+      os,
+      "salt {} server_key {} stored_key {} iterations {}",
+      cred._salt,
+      cred._server_key,
+      cred._stored_key,
+      cred._iterations);
+    return os;
+}
+
+client_first_message::client_first_message(bytes_view data) {
+    static const std::regex re(
```

Comment:
it might be useful for this use case to look at bringing in `ctre` library. I expect this validation to be heavily used? 

so we might want to optimize. std::regex is very slow.

Reply:
got ctre working with these regex formulations.